### PR TITLE
Implementation of the HKDF derivation function

### PR DESCRIPTION
**Pull request by jrihtarsic**
Purpose of the PR is to implement support for the HMAC-based Extract-and-Expand Key Derivation Function (HKDF [[RFC5869](https://datatracker.ietf.org/doc/html/rfc5869)]) used by the KeyAgreement method.
The details of the PR  are in the ticket [SANTUARIO-607](https://issues.apache.org/jira/projects/SANTUARIO/issues/SANTUARIO-607)

The code is contributed on behalf of the European Commission’s edelivery project to support [eDelivery AS4 2.0 profile](https://ec.europa.eu/digital-building-blocks/wikis/pages/viewpage.action?pageId=467117621).

### CODE REVIEW

**Comment by jrihtarsic** [comment]
Hi @coheigea I want to inform you that the update for HKDF xsd scheme for RFC9231  
https://datatracker.ietf.org/doc/draft-eastlake-rfc9231bis-xmlsec-uris/
was accepted and the PR can be reviewed now.

File: src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java
Comment (coheigea) [inline src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java:341]: I guess we should add javadoc here to align with the concat static method?

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFParamsImpl.java
Comment (coheigea) [inline src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFParamsImpl.java:41]: Remove blank line

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFParamsImpl.java
Comment (coheigea) [inline src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFParamsImpl.java:64]: Remove blank lines

File: src/test/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFTest.java
Comment (coheigea) [inline src/test/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFTest.java:107]: Why can't we use the method in XMLCIpherUtil that was added instead of this?

File: src/test/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFTest.java
Comment (coheigea) [inline src/test/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFTest.java:96]: Add space after the = sign

**Review by coheigea — CHANGES_REQUESTED** [review]
@seanjmullan Can you have a look as well please?

File: src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java:339]: Done

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFParamsImpl.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFParamsImpl.java]: Done. I also remove other "double blank lines" in the code.

File: src/test/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFTest.java
Comment (jrihtarsic) [reply src/test/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDFTest.java]: Good catch. Updated and reused the method for tests as well.

**Comment by coheigea** [comment]
@jrihtarsic I'm going to call a vote on 4.0.2/3.0.4 to get the Key Agreement feature released and unblock the next CXF release. The HKDF will have to wait until the next release.

**Comment by jrihtarsic** [comment]
@coheigea understand and thanks for the info. I look forward to the first version of the key agreement option as part of the cxf.

**Comment by seanjmullan** [comment]
> @seanjmullan Can you have a look as well please?

I can try to look at this over the next couple of weeks. I was off work for a while so I am still catching up with things.

**Comment by coheigea** [comment]
@seanjmullan Just a reminder please, as this PR blocks another Pr in WSS4J

**Comment by coheigea** [comment]
@jrihtarsic Can you remove the whitespace changes in this PR? It makes it difficult to get to the actual changes

**Comment by jrihtarsic** [comment]
I reverted the "cleaning of the code" (empty lines) from commit [PR updates (docs and clean empty lines)](https://github.com/apache/santuario-xml-security-java/pull/271/commits/491a7d36b69837b8db2b8192be5af117984ebb68)

File: src/main/java/org/apache/xml/security/resource/xmlsecurity_en.properties
Comment (seanjmullan) [inline src/main/java/org/apache/xml/security/resource/xmlsecurity_en.properties:114]: Nit: most of the other error messages end in a period, so suggest removing it.

File: src/main/java/org/apache/xml/security/resource/xmlsecurity_en.properties
Comment (seanjmullan) [inline src/main/java/org/apache/xml/security/resource/xmlsecurity_en.properties:116]: s/Key/key
s/Algorithm:/algorithm (no need for colon)

File: src/main/java/org/apache/xml/security/encryption/KeyDerivationMethod.java
Comment (seanjmullan) [inline src/main/java/org/apache/xml/security/encryption/KeyDerivationMethod.java:56]: Nit: no need for period.

File: src/main/java/org/apache/xml/security/encryption/params/HKDFParams.java
Comment (seanjmullan) [inline src/main/java/org/apache/xml/security/encryption/params/HKDFParams.java:26]: s/HMacKeyDerivationParameter/HKDFParams/

Try to break lines at around 80 chars.

**Review by seanjmullan — COMMENTED** [review]
I reviewed some of it, so far just a few minor comments. Will need another day or two to finish reviewing.

**Comment by jrihtarsic** [comment]
@seanjmullan  Thank you for checking it. And no worries about time. I'd rather see my code thoroughly vetted by a security expert than to skip/miss some security issues or bugs. So take as much time as you need.

File: src/main/java/org/apache/xml/security/resource/xmlsecurity_en.properties
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/resource/xmlsecurity_en.properties]: Done

**Review by seanjmullan — COMMENTED** [review]
Still reviewing, but here are some more comments.

File: src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java:143]: Totally agree with the comment.  But this wold change the existing behavior and mya break some "useges" which is following the 
https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP
`The message digest function SHOULD be specified using the Algorithm attribute of the ds:DigestMethod child element of the xenc:EncryptionMethod element. If it is not specified, the default value of SHA1 is to be used. `
@coheigea if you agree I will change this to SHA-256,  and I suggest that this is duly noted in the release notes.

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java
Comment (seanjmullan) [inline src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java:78]: Why is `keyLength` a long instead of an int? Do we really think derived keys will be that large?

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java
Comment (seanjmullan) [inline src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java:61]: Why is salt specified here but not in the deriveKey method? This means this HKDF object cannot be reused if salt is different.

I think part of the problem is that the `DerivationAlgorithm` interface is too specific and designed to support ConcatKDF. It may make sense to add some more `deriveKey` methods to that interface that take a `KDFParams` object so that the parameters can be more abstract and defined by the subclasses, ex:

`byte[] deriveKey(byte[] secret, KDFParams params) throws XMLSecurityException`

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java
Comment (seanjmullan) [inline src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java:132]: s/ger/get/

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java
Comment (seanjmullan) [inline src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java:169]: s/secret_key/secretKey/

File: src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java
Comment (seanjmullan) [reply src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java:143]: I think the XML Encryption recommendation made a mistake to allow defaults for algorithms, but I don't think we need to necessarily adhere to that. That said, this could introduce interop issues if we treat a missing DigestMethod as SHA-256 instead of SHA-1. Instead, we could require the application to always specify the MGF digest method and no longer allow null to be passed in as the parameter. 

This may be best handled as a separate issue so that all APIs which have defaults are checked - for example, RSA-OAEP also defaults to SHA-1 for the digest algorithm.

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java:78]: Indeed it is hard to imagine key size to need  long  type. But the interface for DerivationAlgorithm was taken from  ws-wss4 project 
https://github.com/apache/ws-wss4j/blob/master/ws-security-common/src/main/java/org/apache/wss4j/common/derivedKey/DerivationAlgorithm.java
The idea behind was to reduce duplication of the code there (wss4j) and use methods from xmlsec ..  to make wss4j slimmer/cleaner 
Anyhow will change to int.

File: src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java:143]: Created a ticket [SANTUARIO-618](https://issues.apache.org/jira/browse/SANTUARIO-618)

File: src/main/java/org/apache/xml/security/encryption/KeyDerivationMethod.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/KeyDerivationMethod.java:30]: Indeed it was a bit philosophical :) I updated the class description. Hope it's more readable now.

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/KeyDerivationMethodImpl.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/KeyDerivationMethodImpl.java:99]: Fixed the behavior of the method - to match the description.

File: src/main/java/org/apache/xml/security/utils/KeyUtils.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/utils/KeyUtils.java:250]: Good catch, I updated the description

File: src/main/java/org/apache/xml/security/encryption/params/HKDFParams.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/params/HKDFParams.java:25]: done

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java:78]: The interface is changed to have a more generic method as suggested in another comment. I will update the WSS4J  to use this interface as it can be more easily extended with new key derivation parameters.

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java:61]: Done

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/KeyDerivationMethodImpl.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/KeyDerivationMethodImpl.java:102]: Done

**Review by seanjmullan — APPROVED** [review]
Updates looks good.

File: src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java]: I updated the code now to ensure that only the XMLEncryptionException can be thrown. I made changes to the XMLCipherUtil.constructKeyDerivationParameter method only and retained the XMLSecurityException in the interface KeyDerivationMethod.getKDFParams; because the interface implementation may be used for key derivation functions used for other purposes than just the encryption in the future.

File: src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java
Comment (seanjmullan) [inline src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java:277]: typo: "throwen", but I would just remove "thrown in case" as that is implied. Also, a nit but you don't need to end with a period.

Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/XMLCipherUtil.java]: Thanks for the warning about the typo, I fixed the description now.

**Comment by coheigea** [comment]
@seanjmullan Is it ready to be merged from your PoV?

**Comment by coheigea** [comment]
@jrihtarsic There is a build error:
```
[INFO] -------------------------------------------------------------
Error:  COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
Error:  /home/runner/work/santuario-xml-security-java/santuario-xml-security-java/src/test/java/org/apache/xml/security/test/dom/encryption/XMLCipherTest.java:[581,58] ConcatKDFParams(int,java.lang.String) has protected access in org.apache.xml.security.encryption.params.ConcatKDFParams
Error:  /home/runner/work/santuario-xml-security-java/santuario-xml-security-java/src/test/java/org/apache/xml/security/test/dom/encryption/XMLEncryption11BrainpoolTest.java:[123,58] ConcatKDFParams(int,java.lang.String) has protected access in org.apache.xml.security.encryption.params.ConcatKDFParams
```

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java:174]: @coheigea, @seanjmullan, I would need your advice on this check failure. The  method  "XMLCipherUtil.getJCEMacHashForUri" is just mapping URI names to JCE names. Should I just log warning for HmacSHA1, HmacSHA224, HmacSHA384, HMACRIPEMD160 or should the method throw and error since algorithms are weak or should we just ignore the CodeQL?

**Comment by jrihtarsic** [comment]
The branch is now updated with latest changes from the main, the build after the merge should pass now.

File: src/main/java/org/apache/xml/security/utils/KeyUtils.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/utils/KeyUtils.java]: Method renamed

File: src/main/java/org/apache/xml/security/utils/KeyUtils.java
Comment (jrihtarsic) [reply src/main/java/org/apache/xml/security/utils/KeyUtils.java]: Method renamed

File: src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java
Comment (seanjmullan) [reply src/main/java/org/apache/xml/security/encryption/keys/content/derivedKey/HKDF.java:174]: A few comments/questions:

- Why are all of these considered weak/insecure? Ok, I understand HmacSHA1 could be considered weak, but why HmacSHA224, HmacSHA384?
- There is already a secureValidation mode implemented in Santuario which restricts weak crypto algorithms; however HmacSHA1 is not disabled yet. We could consider disabling it, but we also need to assess the compatibility risk of doing that, which may be too high. 
- There may be some more places we should be adding additional secureValidation checks in the HKDF code. Can you explore this further?

**Comment by jrihtarsic** [comment]
> > @seanjmullan Is it ready to be merged from your PoV?
> 
> Yes, although I think we should try to add the secureValidation mode support before we post the next release.
@seanjmullan  I can make the PR for this by the end of the next week. 
The scope is shortly described here: SANTUARIO-620, please let me know if I should implement anything else.

**Comment by coheigea** [comment]
@jrihtarsic Do you need this merged to 3.0.x as well?

**Comment by jrihtarsic** [comment]
> @jrihtarsic Do you need this merged to 3.0.x as well?

@coheigea, yes indeed we would  need it in 3.0.x so that we can use the latest feature with current apache/cxf

**Review by coheigea — APPROVED** [review]

**Pull request merged**
